crates/pico_macros/src/memo.rs

```
@@ -97,56 +90,97 @@ pub(crate) fn memo(args: TokenStream, item: TokenStream) -> TokenStream {
+        .enumerate()
+        .map(|(i, (arg, ty))| {
+            match ArgType::parse(ty) {
+                ArgType::Source => {
+                    let maybe_ref = if matches!(**ty, syn::Type::Reference(_)) {
+                        quote! { &param_id.into() }
+                    } else {
+                        quote! { param_id.into() }
+                    };
+                    quote! {
+                        let #arg: #ty = {
+                            let param_id = derived_node_id.params[#i];
+                            #maybe_ref
+                        };
+                    }
+                }
+                ArgType::MemoRef => {
+                    let maybe_ref = if matches!(**ty, syn::Type::Reference(_)) {
+                        quote! { &::pico::MemoRef::new(#db_arg, param_id.into()) }
+                    } else {
+                        quote! { ::pico::MemoRef::new(#db_arg, param_id.into()) }
+                    };
+                    quote! {
+                        let #arg: #ty = {
+                            let param_id = derived_node_id.params[#i];
+                            #maybe_ref
+                        };
+                    }
+                }
+                ArgType::Other => {
+                    let target_type = if let syn::Type::Reference(ref reference) = **ty {
+                        &reference.elem
+                    } else {
+                        ty
+                    };
+                    let binding_expr = match **ty {
+                        syn::Type::Reference(_) => quote!(inner),
+                        _ => quote!(inner.clone()),
+                    };
+                    quote! {
+                        let #arg = {
+                            let param_ref = ::pico::macro_fns::get_param(#db_arg, derived_node_id.params[#i])?;
+                            let inner = param_ref.downcast_ref::<#target_type>()?;
```

We should probably panic here, instead of returning `None`